File: leetcode_go/question/_101_200/_104/sol_104.go
// Package _104 二叉树的最大深度
// @author goinhn
// @data 2022/8/20
package _104

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 深度优先搜索
// time:O(n)
// space:O(n)
func maxDepth1(root *TreeNode) (count int) {
	var dfsFunc func(root *TreeNode, level int)
	dfsFunc = func(root *TreeNode, level int) {
		if root == nil {
			return
		}

		if count < level {
			count = level
		}

		dfsFunc(root.Left, level+1)
		dfsFunc(root.Right, level+1)
	}

	dfsFunc(root, 1)
	return count
}

// 广度优先搜索
// time:O(n)
// space:O(n)
func maxDepth2(root *TreeNode) (count int) {
	if root == nil {
		return count
	}

	var queue []*TreeNode
	queue = append(queue, root)
	for len(queue) > 0 {
		sz := len(queue)
		for sz > 0 {
			node := queue[0]
			queue = queue[1:]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}

			sz--
		}
		count++
	}

	return count
}

File: leetcode_java/java/com/goinhn/thorwcode/leetcode/question/_101_200/Sol104.java
package com.goinhn.thorwcode.leetcode.question._101_200;

/**
 * 104.二叉树的最大深度
 *
 * @author goinhn
 * @date 2021/11/20
 */
public class Sol104 {

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    /**
     * DFS
     * <p>
     * time:O(n)
     * space:O(n)
     */
    private static class Solution1 {
        public int maxDepth(TreeNode root) {
            return maxSingle(root, 0);
        }

        private int maxSingle(TreeNode root, int deep) {
            if (root == null) {
                return deep;
            } else {
                return Math.max(maxSingle(root.left, deep + 1), maxSingle(root.right, deep + 1));
            }
        }
    }

}
